src/catchup/CatchupWork.cpp

```diff
@@ -0,0 +1,498 @@
+    // can be applied, it is always greater than LCL
+    auto smallestCheckpointToApply =
+        checkpointContainingLedger(smallestLedgerToApply, historyManager);
+
+    // if first ledger that should be applied is on checkpoint boundary then
+    // we do an bucket-apply
+    if (smallestLedgerToApply == smallestCheckpointToApply)
+    {
+        return std::make_pair(true, smallestCheckpointToApply);
+    }
+
+    // we are before first checkpoint - applying buckets is not possible, so
+    // just apply transactions
+    if (smallestCheckpointToApply <= historyManager.getCheckpointFrequency())
+    {
+        return std::make_pair(false, smallestCheckpointToApply);
+    }
+
+    // we need to apply on previous checkpoint (if possible), so we are sure
+    // that we get required number of history entries in database
+    smallestCheckpointToApply -= historyManager.getCheckpointFrequency();
+    if (smallestCheckpointToApply >= smallestBucketApplyCheckpoint)
+    {
+        return std::make_pair(true, smallestCheckpointToApply);
+    }
+    else
+    {
+        // in that case we would apply before LCL, so there is no need to
+        // apply buckets
+        return std::make_pair(false, smallestBucketApplyCheckpoint);
+    }
+}
+}
+
+CatchupRange
+CatchupWork::makeCatchupRange(uint32_t lastClosedLedger,
+                              CatchupConfiguration const& configuration,
+                              HistoryManager const& historyManager)
+{
+    assert(lastClosedLedger > 0);
+    assert(configuration.toLedger() >= lastClosedLedger);
+    assert(configuration.toLedger() != CatchupConfiguration::CURRENT);
+
+    // smallest ledger that should be applied is either firstNeededLedger or
+    // LCL
+    // first variant ensures that we have at least configuration.count() new
+    // history entries in database, second one is selected only when first
+    // one is impossible (as it would require going past LCL again)
+    auto smallestLedgerToApply =
+        std::max(firstNeededLedger(configuration), lastClosedLedger);
+
+    // smallest checkpoint value that can be bucket-applied on local ledger -
+    // all lower checkpoints are already fully applied
+    auto smallestBucketApplyCheckpoint =
+        firstNotFinishedCheckpoint(lastClosedLedger, historyManager);
+
+    // check if catchup should start with bucket apply or not and if so,
+    // which checkpoint should it start at
+    auto catchupStart = computeCatchupStart(
+        smallestLedgerToApply, smallestBucketApplyCheckpoint, historyManager);
+
+    // if we are about to apply buckets just after LCL, we can as well apply
+    // transactions
+    if (catchupStart.first && catchupStart.second <= lastClosedLedger + 1)
+    {
+        catchupStart = std::make_pair(false, catchupStart.second);
+    }
+
+    auto ledgerRange =
+        LedgerRange{std::min(configuration.toLedger(), catchupStart.second),
```

Comment:
when would we have `configuration.toLedger() >= catchupStart.second` ? and when it happens, why would we preserve the value of `catchupStart.first` ?